### ISSUE TITLE
Optimizing

### ISSUE THREAD

**Issue by moham96** [moham96/dtorrent_task#14]
Currently the program is using a bit too much CPU due to multiple reasons:

1- we should try to avoid the use of collection functions since most of them has O(n) complexity, this needs quite a bit of research and testing since it may require testing alternative data structures or maybe roll our own data structures.
2- we should reduce the amount of I/O usage and save the piece in memory before writing to file, validating the piece before that.

3- we should investigate the possibility of caching the files in memory because reading a block each time we want to upload is expensive, maybe we need to use `mmap` which is not possible out of the box in dart right now

**Comment by KabaDH** [comment]
@moham96 hello! Were any of this solved in 0.4.0?
According to the Changelog: "cache piece in memory until it is validated then write to disk"

**Comment by moham96** [comment]
@KabaDH yeah i forgot to check these, just marked the completed tasks (although the piece caching introduced another problem #8), thanks for reminding me
